### DISCUSSION TITLE
Lavalink Node `connection refused`

### DISCUSSION THREAD

**Discussion by RazvanBerbece** [disgoorg/disgolink#55] [Q&A] [answered]
Hello.

I am trying to setup the DiscordGo example that's provided however I'm running into an issue when trying to run the following code:
```
b.Lavalink.AddNode(ctx, disgolink.NodeConfig{
		Name:     NodeName,
		Address:  NodeAddress,
		Password: NodePassword,
		Secure:   NodeSecure,
	})
```

### Stack Trace
```
2023/12/05 03:22:51 ERROR failed to reconnect node name=disgolink_client name=disgolink_node node_name=main err="dial tcp 127.0.0.1:2333: connect: connection refused" try=0
2023/12/05 03:22:53 ERROR failed to reconnect node name=disgolink_client name=disgolink_node node_name=main err="dial tcp 127.0.0.1:2333: connect: connection refused" try=1
2023/12/05 03:22:57 ERROR failed to reconnect node name=disgolink_client name=disgolink_node node_name=main err="dial tcp 127.0.0.1:2333: connect: connection refused" try=2
panic: context deadline exceeded

goroutine 1 [running]:
main.(*Bot).AddLavalinkNode(0xc00009d020, {0x1405ee8, 0xc00011cb60})
        /internal/bot/disgolink.go:42 +0xa31
main.main()
        /internal/bot/main.go:57 +0x2d1
```

I'm not sure how this should work but should AddNode start another service on the host using those params ? It seems like it's not creating anything and that it can't connect ? Unless I'm not understanding how the lavalink stuff behind this works or I'm doing something wrong 🤔 

Can somebody advise ?

Thank you :)

**Comment by topi314** [comment, marked as answer]
You have to host Lavalink yourself https://github.com/lavalink-devs/Lavalink
https://lavalink.dev/getting-started/index.html

**Reply by RazvanBerbece** [reply]
Yeppp just realised, thank you !

I am now trying to play from a designated URL at bot ready time (I'm using DiscordGo), however i'm encountering some issues with these two lines (128, 129)
```
	player := b.Lavalink.Player(snowflake.MustParse(GuildId))
	queue := b.Queues.Get(GuildId)
```

Stack Trace
```
2023-12-05 22:28:23 panic: runtime error: invalid memory address or nil pointer dereference
2023-12-05 22:28:23 [signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x29424c]
2023-12-05 22:28:23 
2023-12-05 22:28:23 goroutine 38 [running]:
2023-12-05 22:28:23 main.(*Bot).onReady(0x5c53a0, 0x1?, 0x3dad78?)
2023-12-05 22:28:23     /app/internal/bot/main.go:128 +0xec
2023-12-05 22:28:23 github.com/bwmarrin/discordgo.readyEventHandler.Handle(0x40000261e0?, 0x0?, {0x2c0660?, 0x40003896e0?})
2023-12-05 22:28:23     /go/pkg/mod/github.com/bwmarrin/discordgo@v0.27.1/eventhandlers.go:948 +0x44
2023-12-05 22:28:23 created by github.com/bwmarrin/discordgo.(*Session).handle in goroutine 1
2023-12-05 22:28:23     /go/pkg/mod/github.com/bwmarrin/discordgo@v0.27.1/event.go:171 +0x14c
```

The `GuildId` that I pass in is a string which contains the server ID obtained from the Discord GUI. I double checked that the variable is populated and it is, so I'm unsure as to why `b.Lavalink.Player(snowflake.MustParse(GuildId))` would fail 🤔  Are there any examples on how to do this outside of a `InteractionCreate` context ?

**Reply by topi314** [reply]
Are you sure `b.Lavalink` is populated?
Or even `b` might be `nil`
I recommend debugging & stepping into the code

**Reply by RazvanBerbece** [reply]
Hey @topi314 thank you for your responses so far ! <3

I have just checked now and it seems that `b.Lavalink` is `nil` (tried to print `b.Lavalink` itself and I got the same error and `<nil>` is being printed. However I am a bit confused as to why it wouldn't work - after all, the only bits which handle setting up `b.Lavalink` can be put into this method

```
func (b *Bot) AddLavalinkNode(ctx context.Context) {
	b.Lavalink = disgolink.New(snowflake.MustParse(b.Session.State.User.ID),
		disgolink.WithListenerFunc(b.onPlayerPause),
		disgolink.WithListenerFunc(b.onPlayerResume),
		disgolink.WithListenerFunc(b.onTrackStart),
		disgolink.WithListenerFunc(b.onTrackEnd),
		disgolink.WithListenerFunc(b.onTrackException),
		disgolink.WithListenerFunc(b.onTrackStuck),
		disgolink.WithListenerFunc(b.onWebSocketClosed),
	)

	node, err := b.Lavalink.AddNode(ctx, disgolink.NodeConfig{
		Name:     NodeName,
		Address:  NodeAddress,
		Password: NodePassword,
		Secure:   NodeSecure,
	})
	if err != nil {
		panic(err)
	}
	version, err := node.Version(ctx)
	if err != nil {
		slog.Error("failed to get node version", slog.Any("err", err))
		os.Exit(1)
	}
	log.Printf("lavalink node version: %s", version)
}
```

**Reply by topi314** [reply]
since I don't have access your code I can only guess.
but my assumption is that you call `player := b.Lavalink.Player(snowflake.MustParse(GuildId))` in the ready event
but you init lavalink after the ready event
after you called `session.Open()` which would cause this issue.
Lavalink needs to be initialized before you call `session.Open()`

**Reply by RazvanBerbece** [reply]
You're actually a genius lol, yep that did it and that was the cause of the issue. Thank you I super appreciate it :) till later
